model_interfaces/llm_interface.py

```diff
@@ -59,25 +101,11 @@ def _ts_message(m: dict) -> dict:
         return context
 
     def reply(self, user_message: str) -> str:
-        context = self.build_context(user_message)
-        if self.verbose:
-            print(f"USER: {user_message}")
-
-        def cost_callback(cost_usd: float):
-            self.costs_usd += cost_usd
-
-        response = ask_llm(
-            context,
-            model=self.model,
-            temperature=0,
-            context_length=self.max_prompt_size,
-            cost_callback=cost_callback,
-        )
-        if self.verbose:
-            print(f"AGENT: {response}")
-        return response
+        self.context = self.build_context(user_message)
+        return super().reply(user_message)
```

Comment:
`self.context` is always overwritten and `self.history` is never updated, so this will have no context at all in practice. Also, the user message will be added twice. What about wrapping the superclass' method?
```python
original_context = self.context
self.context = self.build_context()  # Without user message
response = super().reply(user_message)
self.context = original_context + self.context[-1:]
return response
```

Reply:
I see the need for `self.history`, since you want to keep all the timestamps, but it also needs to be updated.